tools/internal/parser/clean.go

```diff
@@ -0,0 +1,400 @@
+					Key:    "Amazon",
+					Blocks: []Block{block},
+				})
+			case inAmazonGroup:
+				last := len(groups) - 1
+				groups[last].Blocks = append(groups[last].Blocks, block)
+				if strings.Contains(v.Text[0], amazonGroupEnd) {
+					// End of Amazon group, go back to normal
+					// behavior.
+					inAmazonGroup = false
+				}
+			default:
+				// Barrier comment
+				groups = append(groups, group{
+					Key:    "",
+					Blocks: []Block{block},
+				})
+			}
+		case *Suffixes:
+			if inAmazonGroup {
+				last := len(groups) - 1
+				groups[last].Blocks = append(groups[last].Blocks, v)
+			} else {
+				groups = append(groups, group{v.Info.Name, []Block{block}})
+			}
+		default:
+			panic("unknown ast node")
+		}
+	}
+
+	// Scan through the groups, looking for barrier comments. Sort the
+	// groups between the barriers, and just check ordering across
+	// barriers.
+	var (
+		errs           []error
+		prevGroupEnd   *Suffixes
+		prevComment    *Comment
+		thisGroupStart int
+	)
+
+	sortAndCheck := func(groups []group) {
+		if len(groups) == 0 {
+			return
+		}
+
+		slices.SortFunc(groups, func(a, b group) int {
+			return compareCommentText(a.Key, b.Key)
+		})
+
+		if prevGroupEnd == nil {
+			// First group.
+			lastGroup := groups[len(groups)-1]
+			prevGroupEnd = lastGroup.Blocks[len(lastGroup.Blocks)-1].(*Suffixes)
+		} else if compareCommentText(prevGroupEnd.Info.Name, groups[0].Key) <= 0 {
+			// Inter-group order is correct.
+			lastGroup := groups[len(groups)-1]
+			prevGroupEnd = lastGroup.Blocks[len(lastGroup.Blocks)-1].(*Suffixes)
+		} else {
+			// Wrong inter-group order. Report error and keep the same
+			// prevGroupEnd, since it's bigger.
+			errs = append(errs, ErrCommentPreventsSectionSort{prevComment.SourceRange})
+		}
+	}
+
+	for i, group := range groups {
+		if group.Key != "" {
+			// Not a boundary, keep going.
+			continue
+		}
+
+		// Found a boundary.
+		sortAndCheck(groups[thisGroupStart:i])
+		prevComment = group.Blocks[0].(*Comment)
+		thisGroupStart = i + 1
+	}
+	if thisGroupStart != len(groups) {
+		sortAndCheck(groups[thisGroupStart:])
+	}
+
+	// Reassemble the new s.Blocks from groups. Note, we must not
+	// reuse the s.Blocks slice, because all the slices in the groups
+	// are referencing the same backing array. If we start overwriting
+	// that array, we might corrupt future groups and end up with a
+	// list that deletes a bunch of suffix blocks and duplicates a
+	// bunch of others in the wrong place.
+	ret := make([]Block, 0, len(s.Blocks))
+	for _, group := range groups {
+		ret = append(ret, group.Blocks...)
+	}
+	s.Blocks = ret
+
+	return errs
+}
+
+func sortSuffixes(s *Suffixes) []error {
+	// Suffix sorting has the same problem as section sorting: inline
+	// comments act as barriers that prevent movement of a suffix
+	// across them (e.g. "all 3rd-level public suffixes come after
+	// this").
+	//
+	// We do the same thing as for section sorting: sort as much as we
```

Comment:
Do the comments stay attached to the other lines?

Your example is
```
foo.example.com
bar.example.com
// non com domains
foo.example.org
*.example.pl
internal.example.rs
```

but what about
```
foo.example.com
// next is our bar!
bar.example.com
abc.example.com
```

In this case it wouldn't be right to sort `abc` before `bar` and keeping the comment on top.

Reply:
Hm, if the ICANN section has both kinds of comments I have a slight preference for the last option: Make the comments which start subgroups have a different format. For now, can you just disable sorting for the ICANN section so that we don't accidentally mess it up at some point? Then we can put whatever else we do in a different PR.

Reply:
Sounds good. The code already skipped all sorting in the ICANN section, so if you're ok with the current logic for the private section (temporarily, while we implement the new comment format), the PR is good as-is. The private section has no inline comments atm so the sorting logic is a no-op that we can change freely.

Reply:
> Hm, if the ICANN section has both kinds of comments I have a slight preference for the last option: Make the comments which start subgroups have a different format. For now, can you just disable sorting for the ICANN section so that we don't accidentally mess it up at some point? Then we can put whatever else we do in a different PR.

For historical context about the diversity (AKA Sloppy-ness) in the ICANN section, the sub-section stuff became necessary in situations where the IANA Delegee would in turn have a delegee (Example: for .uk, Nominet runs the TLD and most of the SLD subspace like .co.uk, but also there are some which are still administratively part of their delegation but run by other entities like JANET).

The we had the .jp namespace with all the prefectures and their specifically defined structures, which needed annotation as defined by JPRS - which needed a slightly different approach.  Those are just two examples, we have had a number of different approaches used;   When I first started cleaning that up, it would be upon the TLD administrator requesting so, and we'd update their section to whatever the style in use at the time was.

There's some opportunity for cleanup there, but the sub-comments within a 'section' are helpful for context for readers to understand the quite diverse namespace approaches in the ccTLD world.

When we do start to look at the ICANN section, one thing we might want to do is engage with people from their Office of the CTO (OCTO) team and or coordinate with the IANA about it.

That is a much much larger topic with much much larger nuance, and it should be done with elegance (scalpel vs cudgel)